feat(shada): shada should not store nobuflisted buffers

LhKipp opened the pull request:
Problem:  Shada jumplist entries still include entries from e.g. 'nobuflisted' buffers.
Solution: Check `ignore_buf()` before adding jumplist entries, followup to b98eefd8.

Fix #18433

LhKipp commented on runtime/doc/options.txt:
```diff
@@ -5203,9 +5203,8 @@ A jump table for the options with a short description can be found at |Q_op|.
-		('buftype'), unlisted ('buflisted'), unnamed and buffers on 
-		removable media (|shada-r|) are not saved.
+		buffer list is restored from the shada file. Quickfix ('buftype'), 
+		unnamed and buffers on removable media (|shada-r|) are not saved.
```
Removed `unlisted ('buflisted')` from `shada-%`, as `nobuflisted` files are not stored in general.

justinmk replied:
Still worth mentioning explicitly I think. Let's keep the help as it was here.

LhKipp replied:
Okay, lets do that. 

The change is reverted by: e88fd1cb512d525d83dde3a1113335e9785f777d

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5203,9 +5203,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	%	When included, save and restore the buffer list.  If Vim is
```
Testing `:set shada+=%`, it looks like it is _already_ the case that `'nobuflisted'` buffers are _not_ saved/restored (check with `:ls!`). So the change in this PR actually just fixes a bug, maybe? 

Anyway, let's update the doc here:
```suggestion
	%	Save and restore the buffer list (only 'buflisted' buffers).  If Vim is
```

justinmk commented on runtime/doc/starting.txt:
```diff
@@ -920,7 +920,7 @@ The ShaDa file is used to store:
 - File marks, pointing to locations in files.
 - Last search/substitute pattern (for 'n' and '&').
-- The buffer list.
+- The buffer list (without |nobuflisted| buffers).
```
`|nobuflisted|` is not a valid tag, but anyway let's change this to:
```suggestion
- The buffer list (only 'buflisted' buffers).
```

LhKipp replied:
Thanks. Makes sense to remove the double negation

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -262,4 +262,14 @@ describe('ShaDa support code', function()
+    curbufmeths.set_option('buflisted', false)
+    nvim_command('wshada! ' .. shada_fname)
+
+    eq({}, find_file(fname))
```
to avoid false negatives (tests that pass but shouldn't), it helps to check the reverse, so e.g. after this line add a small sanity check:


    curbufmeths.set_option('buflisted', true)
    nvim_command('wshada! ' .. shada_fname)
    eq(..., find_file(fname))

LhKipp replied:
Makes sense.

e12dee2440177cb4930ff3e8912aeb74860d8fd8

justinmk reviewed (approved):
lgtm. if there is no feedback from others, ping me if a few days to merge.

thanks!

LhKipp commented:
@justinmk Hi Justin, thanks for your review. Pinging you in order to get this merged :) 

I've force pushed, as there has been a conflict in news.txt.

justinmk commented:
Looks like the test fails on windows?

```
FAILED   test/functional\shada\shada_spec.lua @ 266: ShaDa support code does not store unlisted buffer
test/functional\shada\shada_spec.lua:278: Expected objects to be the same.
Passed in:
(table: 0x01f097385780) { }
Expected:
(table: 0x01f0973856d0) {
 *[7] = 1
  [8] = 1
  [10] = 1 }

stack traceback:
	test/functional\shada\shada_spec.lua:278: in function <test/functional\shada\shada_spec.lua:266>
```

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -262,4 +262,19 @@ describe('ShaDa support code', function()
+    -- Set 'buflisted', then check again.
+    curbufmeths.set_option('buflisted', true)
+    nvim_command('wshada! ' .. shada_fname)
+    eq({[7]=1, [8]=1, [10]=1}, find_file(fname))
```
because this depends on the filesystem it can be racy. try wrapping this in `retry()` 
```suggestion
    retry(nil, 4000, function()
      eq({[7]=1, [8]=1, [10]=1}, find_file(fname)) end)
```

justinmk commented:
windows CI is still failing. what's different about this test compared to others in that test file?

```
test/functional\shada\shada_spec.lua:278: retry() attempts: 133
test/functional\shada\shada_spec.lua:279: Expected objects to be the same.
Passed in:
(table: 0x02266e573058) { }
Expected:
(table: 0x02266e572fa8) {
 *[7] = 1
  [8] = 1
  [10] = 1 }

stack traceback:
	test\helpers.lua:64: in function 'retry'
	test/functional\shada\shada_spec.lua:278: in function <test/functional\shada\shada_spec.lua:266>

LhKipp commented:
Hi Justin,

thanks for keeping an eye on this PR. I currently can't find the time to check why this is failing on windows. I'll come back to this probably next week.

I've marked this MR as draft for now. Sorry for all the disturbances.

LhKipp commented:
> windows CI is still failing. what's different about this test compared to others in that test file?

I can't figure out why this test case fails on windows. I've checked, whether there is some os-specific logic in shada.c, but I couldn't find any. I don't have a windows machine to test with and I am kinda clueless here.

I've disabled this test on windows for now. If this is unacceptable, I am happy if anyone picks up this PR and fixes the windows testcase.

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -264,4 +264,21 @@ describe('ShaDa support code', function()
+
+    local fname = funcs.getcwd() .. '/file'
+    meths.set_var('__fname', fname)
+    nvim_command('silent! edit `=__fname`')
```
why silent with "!"? that silences errors...

LhKipp replied:
Removed `silent!`.
I've created this test by mostly coping another one. The other one did a silent write.

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -264,4 +264,21 @@ describe('ShaDa support code', function()
   end)
+
+  it('does not store unlisted buffer', function()
+    skip(is_os('win'))
```
there's no good reason to skip this on windows. we need to figure out why the test isn't working there

LhKipp replied:
Hi Justin,

thanks for looking at the PR again. As mentioned in [this comment](https://github.com/neovim/neovim/pull/21818#issuecomment-1414145819) I've disabled the test, because I couldn't figure out why the test is failing on windows.

I wanted to try, whether the test would succeed on windows after rebasing on the latest master commit - it doesn't.

I am sorry for taking your time and not properly marking this PR as Draft. ✌🏻

luukvbaal committed:
fix(shada): ignore buffers for jumplist

Problem:  Shada jumplist entries still include entries from e.g. 'nobuflisted' buffers.
Solution: Check `ignore_buf()` before adding jumplist entries, followup to b98eefd8.

Co-authored-by: Leonhard Kipp <Leonhard.Kipp@alumni.fh-aachen.de>
